src/java/org/apache/cassandra/index/sai/disk/v1/kdtree/BKDReader.java

```diff
@@ -880,4 +780,169 @@ public interface IntersectVisitor
+        // time, we know it's done, and we have to go up at least one more level.
+        //
+        // Note that we're storing levels, because we're interested only in the nodes on the current path from
+        // the root of the tree, as those are the only nodes that could be explored. A more obvious alternative
+        // would be to keep a set of all already visited node ids in the tree, but that would have worse memory
+        // complexity and would likely require heap allocations.
+        private final BitSet completedLevels;
+
+        /**
+         * Creates the cursor over the KD-tree leaves and positions it on the first leaf
+         * appropriate for the given query and traversal direction.
+         * Even if the query does not match any data, the cursor is positioned on one of the tree leaves,
+         * so {@link #getFilePointer()} and {@link #getNodeId()} can be always called immediately after the construction.
+         *
+         * @param query restricts the leaves to the ones that might contain the data that match the query,
+         *              null query means the range is not restricted
+         */
+        LeafCursor(Direction direction, @Nullable IntersectVisitor query)
+        {
+            this.query = query;
+            this.direction = direction;
+
+            completedLevels = new BitSet(64); // physically impossible to have a tree bigger than 2^64 nodes
+            tree = new PackedIndexTree();  // this positions the tree at node id 1 (not 0)
+
+            if (direction == Direction.FORWARD)
+                pushToMinLeaf(query);
+            else
+                pushToMaxLeaf(query);
+        }
+
+        /**
+         * Returns the id of the node the cursor is positioned at.
+         * Valid only immediately after construction or after a call to {@link #advance()} which returned {@code true}.
+         */
+        int getNodeId()
+        {
+            assert tree.isLeafNode();
+            return tree.nodeID;
+        }
+
+        /**
+         * Returns the file pointer of the node the cursor is positioned at
+         * Valid only immediately after construction or after a call to {@link #advance()} which returned {@code true}.
+         */
+        long getFilePointer()
+        {
+            assert tree.isLeafNode();
+            return tree.getLeafBlockFP();
+        }
+
+        /**
+         * Advances the cursor to the next leaf.
+         * If there are no more leaves in the tree at all, positions the index tree at node 0.
+         * If there exist leaves, but they are out of the query range, positions the index tree at a non-leaf node.
+         *
+         * @return true if the cursor was moved to the next leaf, false if there are no more leaves to iterate
+         */
+        boolean advance()
+        {
+            completedLevels.set(tree.level);
+
+            if (!popToFirstUncompletedLevel())
+                return false;
+
+            completedLevels.set(tree.level);
```

Comment:
Can we remove this call?

Reply:
No. This one is extremely important. This marks a different level, than the one we marked before `popToFirstUncompletedLevel()`. We must mark it as completed, because we're reaching this level for the second time now (the first time we reached it coming down from the top; now we're returning to it from the bottom), so this means we already visited one child and we're going to be visiting the second one. We must mark it, so when we return to this very same place again in `popToFirstUncompletedLevel()` it would be skipped over and we'll jump up to a level above it. 

I'll add a comment explaining this.

Reply:
I found the way to even improve code more. Instead of separately marking the completeness here, now I moved it to the `pushXXX` methods, so it can't be accidentally forgotten. This way, whenever we go down the tree, we keep the class invariant that all levels above the current level have always a proper completeness status. So this line no longer exists here, but the logic is still there, only in different place. It also allowed to remove similar line in `pushToLeaf`.